### ISSUE TITLE
When copying and pasting an entry to another bib file - all attached files should also be copied

### ISSUE THREAD

**Issue by koppor** [JabRef/jabref#12267]
As user, I have my `personal.bib` and a collaboratively edited `paper.bib` (Refs https://github.com/JabRef/jabref/issues/160). The `paper.bib` has configured a (separate) file directory where the PDFs are stored. If this is true, I would like to have to following behavior:

1. I copy an entry of `personal.bib`. That entry has attached a PDF `A.pdf`.
2. I paste that entry to `paper.bib`.
3. `A.pdf` should be copied to the file directory of library `paper.bib`

In summary: All attached files should be copied if original files are not reachable from new library. One needs to keep the relative path structure while copying (to keep the files field unmodified)

**Comment by Nagarjuna325** [comment]
Hi Koppor, My name is nagarjuna.
I want to work on this issue can you please assign me this task for me.

**Comment by Nagarjuna325** [comment]
I am requesting for 2 more days to finish the work.

**Comment by koppor** [comment]
> I am requesting for 2 more days to finish the work.

Do you have an initial code? Then, please submit a pull request.

Maybe, you know that the work is "finished" if it is merged into the main branch. The code needs to be properly reviewed.

Finally, I am not sure on which base you request the "2 more days". "14 April 2025" plus 2 days has already passed. From your message to today, 2 days have already passed. From the remidner, you have until April, 24, which is more than 2 days you asked three days ago.

---

I am taking no action. Either, there is a PR on April 23 or you will be unassigned.

**Comment by Nagarjuna325** [comment]
Hi Koppor, 

Thanking for giving me an extra week to complete the work. I finished it I will be submitting the PR right now.

**Comment by UmutAkbayin** [comment]
Hi.
I'd like to give this a try. It seems like no one is actively working on it right now, so I'm happy to explore the codebase and see if I can implement the file-copy behavior as described.

I'm new to JabRef.

**Comment by UmutAkbayin** [comment]
I tried the following structure:
.
├── paper
│   └── paper.bib
└── personal
    ├── personal.bib
    └── something.txt

personal.bib has a file field pointing to something.txt. When I copy the entry from personal.bib into paper.bib, I can still see the attachment in the GUI, but trying to open it from within paper.bib fails.

Just to confirm: In this case, the file something.txt should be copied to the paper directory so that it becomes reachable relative to paper.bib, right?

**Comment by koppor** [comment]
@umu directory structure a bit strange. But OK. I am one the road. Short answer: seems right. Please keep the relative path; but the file has to be copied (because the starting point changes). Read on at https://devdocs.jabref.org/code-howtos/#get-a-relative-filename-or-path-for-a-file and think of different cases. Maybe the file is reachable from the other bib file without copying; then just the path needs to be adapted.

**Comment by UmutAkbayin** [comment]
Sorry for asking again, I just want to make sure I fully understand the intended behavior.

Let’s say I have the following structure:

<img width="227" alt="Image" src="https://github.com/user-attachments/assets/a852d334-b47e-4bc0-b402-7cd48e07b330" />

When I copy an entry from personal.bib to parent.bib, then the file is not reachable and needs to be copied, that part is clear.

But if I copy an entry from personal.bib to paper.bib, and paper.bib has @ Comment{jabref-meta: fileDirectory:paperfiles;}, the file is still reachable without copying.

However, the current implementation keeps the file field as:

file = {:personalfiles/something.txt:TXT}

And you would like that to be changed to a relative path like:

file = {:../personalfiles/something.txt:TXT}

So that it works correctly from the perspective of paper.bib with its fileDirectory set to paperfiles/?

Thanks again for your guidance and for your time!

**Comment by koppor** [comment]
> Let’s say I have the following structure:

Note that this structure is strange, because the "bibs" folder is a sub folder of the directory containing the `parent.bib`.

You also need to say HOW the directories are configured for each `.bib` file.

I assume you have `.` configured as directory for each.

![Image](https://github.com/user-attachments/assets/8d8aa826-daac-4ad8-bf20-e03155e046f8)

> When I copy an entry from personal.bib to parent.bib, then the file is not reachable and needs to be copied, that part is clear.

Case 1: Directory reachable.

- File path in personal.bib: `personalfiles/something.txt`
- File path in parent.bib: `bibs/personalfiles/something.txt`

Only path needs to be adapted. From the first to the second.

Case 2: Directory not reachable

- Assume that `personal.bib` has `personalfiles` configured as file directory.
- Assume that `parent.bib` has configured `parentfiles` as file directory.

Then, the relative path is `something.txt` for `personal.bib` - and also for `parent.bib`. Thus no need to adapt anything.

Case 3: Directory not reachable and relative paths differ

- Assume that `personal.bib` has `.` configured as file directory.
- Assume that `parent.bib` has configured `parentfiles` as file directory.

Then, the relative path is `something.txt` for `personal.bib`. It should be `personalfiles/something.txt` for `parant.bib` AND `./parerntfiles/personalfiles`  be created as directory - to keep the nested directory structure. 

> But if I copy an entry from personal.bib to paper.bib, and paper.bib has @ Comment{jabref-meta: fileDirectory:paperfiles;}, the file is still reachable without copying.

Then, your assumption is that `parent.bib` has `bibs/paperfiles` configured and `personal.bib`, `paperfiles`. Then the relative paths starting from that are the same for both.

In case BOTH have `paperfiles` configured, the first path is `./paperfiles`, the second one is `bibs/paperfiels`. But they are DIFFERENT (see the `bib` included in the second). And this is case 1.

> However, the current implementation keeps the file field as:
> 
> file = {:personalfiles/something.txt:TXT}
> 
> And you would like that to be changed to a relative path like:
> 
> file = {:../personalfiles/something.txt:TXT}

In case the directory is "reachable" (based on the directory configuration). If the file is non-reachable, it needs to be copied. Case 3.

You need to document this. Create a file `files.md` in https://github.com/JabRef/jabref/tree/main/docs/requirements. Follow the guidance at `index.md` there how to do.

**Comment by UmutAkbayin** [comment]
Thank you for the detailed explanation! I’ll try to implement and document it over the weekend.

**Comment by UmutAkbayin** [comment]
Hi.

I’ve implemented a working solution for the file copying functionality (via LinkedFileTransferHelper) and integrated it after importHandler.importEntriesWithDuplicateCheck in both CopyTo#copyEntriesWithFeedback and LibraryTab#pasteEntry.

Current behavior:

While the file copying and path adjustment work as expected, I noticed that updating the file links in the imported entries also affects the original entries in the source database. Upon investigation, this seems to be due to the fact that importEntryWithDuplicateCheck doesn’t create a deep copy (the same BibEntry instance is used in both databases).


My understanding:

Only the target entry should be updated during import, not the source one. The observed behavior seems unintentional and potentially problematic.

So the marked line in the screenshot below should be changed in the target, but not the source, correct?

<img width="359" height="251" alt="Image" src="https://github.com/user-attachments/assets/3e5370e1-3c0b-4db8-a559-c2c864f5114c" />

Questions:
	1.	Is this shared BibEntry behavior intentional, or should imported entries be treated as independent copies?
	2.	If independence is the intended behavior, would it be acceptable for me to introduce a copying mechanism (deep or shallow), for example, before cleanup or during the import?

Thanks in advance for the clarification!

**Comment by koppor** [comment]
> Only the target entry should be updated during import, not the source one. 

You have to use .clone(). See other places in the code, e.g. paste entry.

Good that you noticed that!

Background: There is no EMF model in place ensuring proper composition (a BibEntry may belong to one BibDatabase only).

**Comment by UmutAkbayin** [comment]
I’ve opened a PR to address this issue.
It adds the requirement documentation and tests the three file-transfer cases.
Feedback on the business logic, especially the last case with nested paths, would be appreciated.

Sorry for the delay, it was a busy week.

Edit: I see there are issues. I will take care of them next week as good as I can. 
I’ve also run the full test suite in my branch and on main. Three (sometimes four) tests fail in both, so they seem unrelated to my changes.
The bot shows me an issue with BibEntryTest, but this was not touched by me. apparently there was an issue with my rebase.

**Comment by UmutAkbayin** [comment]
Is there anything else I need to do, or is it ready for review? The issues I mentioned above were only related to PR and trag-bot. The code itself is in the final state as I envision it.

**Comment by koppor** [comment]
@UmutAkbayin This popped up at my GitHub notifications and I opened this issue. You commented on an issue, you did not link the PR or commented at the PR. Thus, I could not find the PR. Sorry.

**Comment by koppor** [comment]
@UmutAkbayin Now, I found the PR. Did you check it for yourself? Seems not! Tests failing. Let's continue the discussion there at https://github.com/JabRef/jabref/pull/13535
